### PR TITLE
limit the max size of raft entry

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
If entry exceed the max size, reject to propose it.
@disksing Should we handle the new error(`RaftEntryTooLarge`) in TiClient?

@siddontang @BusyJay @disksing PTAL

### CODE REVIEW

**Comment by disksing** [comment]
Looks like we should handle it particularly. For unknown region errors the default behavior is clear cache and retry.

**Comment by siddontang** [comment]
Please add a test.

File: src/raftstore/store/config.rs [hunk 1]
```diff
@@ -21,6 +21,7 @@ const RAFT_HEARTBEAT_TICKS: usize = 10;
 const RAFT_ELECTION_TIMEOUT_TICKS: usize = 50;
 const RAFT_MAX_SIZE_PER_MSG: u64 = 1024 * 1024;
 const RAFT_MAX_INFLIGHT_MSGS: usize = 256;
+const RAFT_ENTRY_MAX_SIZE: u64 = 16 * 1024 * 1024;
```
Comment (BusyJay) [inline src/raftstore/store/config.rs]: 16M seems too large. I think 4M should be enough.

File: tests/raftstore/test_large_entry.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        let (k, v) = (format!("key{}", i), format!("value{}", i));
+        let key = k.as_bytes();
+        let value = v.as_bytes();
+        cluster.must_put(key, value);
+        let v = cluster.get(key);
+        assert_eq!(v, Some(value.to_vec()));
+    }
+
+    // construct large raft entry
+    let mut large_value: Vec<u8> = Vec::with_capacity(max_size + 1);
```
Comment (BusyJay) [inline tests/raftstore/test_large_entry.rs]: Use `vec!['v'; max_size + 1]`.

File: src/raftstore/store/config.rs [hunk 3]
```diff
@@ -21,6 +21,7 @@ const RAFT_HEARTBEAT_TICKS: usize = 10;
 const RAFT_ELECTION_TIMEOUT_TICKS: usize = 50;
 const RAFT_MAX_SIZE_PER_MSG: u64 = 1024 * 1024;
 const RAFT_MAX_INFLIGHT_MSGS: usize = 256;
+const RAFT_ENTRY_MAX_SIZE: u64 = 8 * 1024 * 1024;
```
Comment (BusyJay) [inline src/raftstore/store/config.rs]: This is inconsistent with `config-template.toml`.

File: tests/raftstore/test_large_entry.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    for i in 1..1000 {
+        let (k, v) = (format!("key{}", i), format!("value{}", i));
+        let key = k.as_bytes();
+        let value = v.as_bytes();
+        cluster.must_put(key, value);
+        let v = cluster.get(key);
+        assert_eq!(v, Some(value.to_vec()));
+    }
+
+    // construct large raft entry
```
Comment (BusyJay) [inline tests/raftstore/test_large_entry.rs]: I prefer add this snippets in `test_single.rs`.

**Comment by BusyJay** [comment]
LGTM

**Comment by zhangjinpeng1987** [comment]
PING

**Comment by siddontang** [comment]
LGTM

**Pull request merged**